man/suck.1
.\" $Revision: 4.2.0 $
.TH SUCK 1
.SH NAME
suck - Pull a small newsfeed from an NNTP server, avoiding the NEWNEWS command.
.SH SYNOPSIS
.I suck
[
.BI
hostname
]
[
.BI @filename
]
[
.BI \-V
]
[
.BI \-K
]
[
.BI \-L[SL]
]
[
.BI \-LF
filename
]
[
.BI \-H
]
[
.BI \-HF
filename
]
[
.BI \-d[tmd]
dirname
]
[
.BI \-s\ |\ \-S
filename
]
[
.BI \-e\ |\ \-E
filename
]
[
.BI \-a
]
[
.BI \-m
]
[
.BI \-b[irlf]
batchfile
]
[
.BI \-r
filesize
]
[
.BI  \-p
extension
]
[
.BI \-U
userid
]
[
.BI \-P
password
]
[
.BI \-Q
]
[
.BI \-c
]
[
.BI \-M
]
[
.BI \-N
port_number
]
[
.BI \-W
pause_time pause_nr_msgs
]
[
.BI \-w
pause_time pause_nr_msgs
]
[
.BI \-l
phrase_file
]
[
.BI \-D
]
[
.BI \-R
]
[
.BI \-q
]
[
.BI \-C
count
]
[
.BI \-k
]
[
.BI \-A
]
[
.BI \-AL
activefile
]
[
.BI \-hl
localhost
]
[
.BI \-bp
]
[
.BI \-T
timeout
]
[
.BI \-n
]
[
.BI \-u
]
[
.BI \-z
]
[
.BI \-x
]
[
.BI \-B
]
[
.BI \-O
]
[
.BI \-G
]
[
.BI \-X
]
[
.BI \-f
]
[
.BI \-y
post_filter
]
[
.BI \-F
]
[
.BI \-g
]
[
.BI \-i
number_to_read
]
[
.BI \-Z
]
[
.BI \-rc
]
[
.BI \-lr
]
[
.BI \-sg
]
[
.BI \-ssl
]
[
.BI \-SSL
]

Options valid in all modes
\hostname

The hostname may optionally include the port number, in the form
.BI Host:Port.
If this option is used, any port number specified
via the -N option is ignored.

\@filename 

This option tells suck to read other options from a file in addition to the
commandline.

\-a

This option forces suck to always batch up any downloaded articles,
even if suck aborts for any reason.  Without this option, suck will
only batch up articles if it finishes successfully or is cancelled by
a signal (see below).

\-A

This option tells suck to scan the localhost (specified with the \-hl option) and use its active file
to build and update the sucknewsrc.   If you add a group to your local server, suck will add it to
sucknewsrc and download articles.  Or, if you delete a group from your local server, it will be deleted
from sucknewsrc.  If posting is not allowed to a particular group, then the line in sucknewsrc is
just commented out.  With this option, you should never have to edit your sucknewsrc.  In case you have
newsgroups (like control and junk) that you don't want downloaded, you can put these newsgroups in a
file "active-ignore", one per line, and suck will ignore these newsgroups when it scans the localhost.
If your system supports regex(), you may use regular expressions in the active-ignore file to skip multiple groups, eg: fred\.*.
If you use the -p (postfix) option, suck will check for the existence of an active-ignore file with the
postfix.  If that doesn't exist, then suck will check for the existence of the file without the postfix.

NOTE: If the localhost is on a non-standard port, the port number may be specified as part of the hostname,
in the form 
.BI Host:Port.

NOTE: If you use regular expressions, suck will silently add a "^" to the beginning of the group name, 
and a "$" to the end of the group name if they aren't already present, so that if you have "comp.os.linux",
it won't match "comp.os.linux.answers" or if you have "alt.test" it doesn't match "comp.alt.test".

\-AL activefile

This option is identical to the -A option, except it reads the active file from the local file specified instead of 
reading it from the localhost.  All the caveats from the -A option apply to this option as well.  If both
options are used on the command line, suck first tries to use the -A option, then if that fails it uses
this option.

\-B

This option tells suck to attempt to batch up any articles in its directory
BEFORE starting to download messages.  This can be useful if you have a 
problem with the previous download.  This option will only work if you specify
a batch option (see below).   If there are no messages to batch up, some
of the batch options may produce warning messages.  They may be safely ignored.
Also, if the batch files exist at the end of the run, in inn-batch mode, it
will be overwritten, since the new batch file will contain all messages.  In
rnews mode, if the batch file exists, it will abort and not batch up any messages.

\-c

If this option is specified, suck will clean up after itself.  This includes:
.RS
1. Moving sucknewsrc to sucknewsrc.old
.RE
.RS
2. Moving suck.newrc to sucknewsrc
.RE
.RS
3. rm suck.sorted and suckothermsgs.
.RE

\-C count

This option tells suck to drop the connection and reopen it every count number of articles.
This is designed to battle INN's LIKE_PULLERS=DONT option, that some folks compile in. With
LIKE_PULLERS=DONT, after 100 messages INN will pause between every message, dramatically 
reducing your download speed. I don't recommend the use of this, but if you have no other choice....  

\-dd dirname

\-dm dirname

\-dt dirname

Specify the location of the various files used by suck.

\-dd dirname = directory of data files used by suck (sucknewsrc suckkillfile suckothermsgs active-ignore sucknodownload)

\-dm dirname = directory for storage of articles created in Multifile mode
or batch mode.  DO NOT make this the same as the directories used for the
\-dt or -\dd options, or you will lose all your configuration files.

\-dt dirname = directory of temp files created by suck (suck.newrc, suck.sort, suck.restart, suck.killlog, suck.post).

\-D

This option tells suck to log various debugging messages to "debug.suck", primarily
for use by the maintainer.

\-e | \-E filename

These options will send all error messages (normally displayed on stderr), to 
an alternate file.  The lower case version, -e, will send the error messages
to the compiled-in default defined in suck_config.h.  The default is suck.errlog.
The upper case version, -E, requires the filename parameter.  All error messages
will then be sent to this file.

\-f 

This option tells suck to reconnect after deduping, and before downloading the articles.  This is in case
long dedupe times cause timeouts on the remote end. 

\-F

This option tells suck to reconnect after reading the local active file, and before downloading the Msg-IDs.
This is in case of a large active file, which causes timeouts on the remote end.

\-g

This option causes suck to only download the headers of any selected articles.
As a result of this, any batching of articles is skipped.   This option does
work with killfiles, however, killfile options such as BODYSIZE> will be
ignored, since the body of the article will never be downloaded.
 
\-G

This option causes suck to display the message count and BPS status lines in a slightly different format,
more suitable for use by a filter program (such as a GUI).

\-H

This option will cause suck to bypass the history check.

\-HF history_file_name

This option tells suck the location of the history file.  The default is at /var/lib/news/history.

\-hl localhost

This option specifies the localhost name.  This option is required with both the \-A and the \-bp option.  

\-i number_to_read

This option tells suck the number of articles to download if you are using the -A
or -AL option, and a new group is added.   The default is defined in suck_config.h (ACTIVE_DEFAULT_LASTREAD, currently -100).  NOTE:  This must be a negative
number (eg -100, -50), or 0, to download all articles currently available in 
the group.

\-k

This option tells suck to NOT attach the postfix from the \-p option to the names of the killfiles,
both the master killfile and any group files.  This allows you to maintain one set of killfiles for
multiple servers.

\-K

This option will cause suck to bypass checking the killfile(s).

\-l phrase_file

This option tells suck to load in an alternate phrase file, instead of using
the built-in messages.  This allows you to have suck print phrases in another
language, or to allow you to customize the messages without re-building suck.
See below.

\-lr 

This option, is used in conjunction with the highest article option in the sucknewsrc, to
download the oldest articles, vice the newest articles. See that section for more details.

\-L

This option tells suck to NOT log killed articles to suck.killlog.

\-LF filename

This option allows you to override the built-in default of "suck.killlog" for the
file which contains the log entries for killed articles.

\-LL

This option tells suck to create long log entries for each killed article.  The long
entry contains the short log entry and the header for the killed message.

\-LS

This option tells suck to create short log entries for each killed article.  The short
entry contains which group and which pattern was matched, as well as the MsgID of the
killed article.

\-M

This option tells suck to send the "mode reader" command to the remote
server.  If you get an invalid command message immediately
after the welcome announcement, then try this option.

\-n

This option tells suck to use the article number vice the MsgId to retrieve the articles.  This
option is supposedly less harsh on the remote server.  It can also eliminate problems if your
ISP ages off articles quickly and you frequently get "article not found" errors.
Also, if your ISP uses DNEWS, you might need this option so that it knows you're reading articles in a group.

\-N port_number

This option tells suck to use an alternate NNRP port number when connecting
to the host, instead of the default, 119.

\-O

This option tells suck to skip the first article upon restart.  This is used whenever
there is a problem with an article on the remote server.  For some reasons, some
NNTP servers, when they have a problem with a particular article, they time out.  
Yet, when you restart, you're back on the same article, and you time out again.
This option tells suck to skip the first article upon restart, so that you can
get the rest of the articles.

\-p extension

This extension is added to all files so that you can have multiple site feeds.
For example, if you specify -p .dummy, then suck looks for sucknewsrc.dummy, suckkillfile.dummy,
etc, and creates its temp files with the same extension.  This will allow you to keep
multiple sucknewsrc files, one for each site.

\-q

This option tells suck to not display the BPS and article count messages during download.
Handy when running suck unattended, such as from a crontab.

\-R

This option tells suck to skip a rescan of the remote newserver upon a restart.  The
default is to rescan the newserver for any new articles whenever suck runs, including
restarts.

\-rc

This option tells suck to change its behavior when the remote server resets its article
counters.   The default behavior is to reset the lastread in sucknewsrc to the current
high article counter.  With this option, suck resets the lastread in sucknewsrc to the
current low article counter, causing it to suck all articles in the group, and using
the historydb routines to dedupe existing articles.

\-s | \-S filename

These options will send all status messages (normally displayed on stdout), to
an alternate file.  The lower case version, -s, will send the status messages
to the compiled-in default defined in suck_config.h.  The default is /dev/null,
so no status messages will be displayed.  The upper case version, -S, requires
the filename parameter.  All status messages will then be sent to this file.

\-sg

This option tells suck to add the name of the current group being downloaded, if known, 
to the BPS display.   Typically the only time suck doesn't know the group name is if
an article is downloaded via the suckothermsgs file.

\-ssl 

This option tells suck to use SSL to talk to the remote server, if suck was compiled with
SSL support.

\-SSL

This option tells suck to use SSL to talk to the local server, if suck was compiled with
SSL support.

\-T timeout

This option overrides the compiled-in TIMEOUT value. This is how long suck waits for data from the
remote host before timing out and aborting.  The timeout value is in seconds.

\-u 

This option tells suck to send the AUTHINFO USER command immediately upon connect to the
remote server, rather than wait for a request for authorization.  You must supply the
\-U and \-P options when you use this option.

\-U userid

\-P password

These two options let you specify a userid and password, if your NNTP server
requires them.

\-Q

This option tells suck to get the userid and password for NNTP authentication from
the environment variables "NNTP_USER" and "NNTP_PASS" vice the -U or -P password.
This prevents a potential security problem where someone doing a ps command can
see your userid and password.

\-V

This option will cause suck to print out the version number and then exit.

\-w pause_timer pause_nr_msgs

This option allows you to slow down suck while pulling articles.  If you
send suck a predefined signal (default SIGUSR1, see suck_config.h),
suck will swap the default pause options (if specified by the -W option),
with the values from this option.  For example, you run suck with -w 2 2,
and you send suck a SIGUSR1 (using kill), suck will then pause 2 seconds
between every other message, allowing the server to "catch its breath."
If you send suck another SIGUSR1, then suck will put back the default
pause options.  If no pause options were specified on the command line
(you omitted -W), then suck will return to the default full speed pull.

\-W pause_time pause_nr_msgs

This option tells suck to pause between the download of articles.  You need
to specify how long to pause (in seconds), and how often to pause (every X nr
of articles). Ex: \-W 10 100 would cause suck to pause for 10 seconds every
100 articles.  Why would you want to do this?  Suck can cause heavy loads on 
a remote server, and this pause allows the server to "catch its breath."

\-x

This option tells suck to not check the Message-IDs for the ending > character.  This option
is for brain dead NNTP servers that truncate the XHDR information at 72 characters.

\-X

This option tells suck to bypass the XOVER killfiles.

\-y post_filter

This option is only valid when using any of batch modes.  It allows you to edit any or all of
the articles downloaded before posting to the local host.   See below for more details.

\-z

This option tells suck to bypass the normal deduping process.  This is primarily for
slow machines where the deduping takes longer than the download of messages would.  Not
recommended.

\-Z

This option tells suck to use the XOVER command vice the XHDR command to retrieve the 
information needed to download articles.  Use this if your remote news server doesn't
support the XHDR command.

.SH LONG OPTION EQUIVALENTS
.RS
\-a  \-\-always_batch
.RE
.RS
\-bi \-\-batch-inn
.RE
.RS
\-br \-\-batch_rnews
.RE
.RS
\-bl \-\-batch_lmove
.RE
.RS
\-bf \-\-batch_innfeed
.RE
.RS
\-bp \-\-batch_post
.RE
.RS
\-c  \-\-cleanup
.RE
.RS
\-dt \-\-dir_temp
.RE
.RS
\-dd \-\-dir_data
.RE
.RS
\-dm \-\-dir_msgs
.RE
.RS
\-e  \-\-def_error_log
.RE
.RS
\-f  \-\-reconnect_dedupe
.RE
.RS
\-g  \-\-header_only
.RE
.RS
\-h  \-\-host
.RE
.RS
\-hl \-\-localhost
.RE
.RS
\-k  \-\-kill_no_postfix
.RE
.RS
\-l  \-\-language_file
.RE
.RS
\-lr \-\-low_read
.RE
.RS
\-m  \-\-multifile
.RE
.RS
\-n  \-\-number_mode
.RE
.RS 
\-p  \-\-postfix
.RE
.RS
\-q  \-\-quiet
.RE
.RS
\-r  \-\-rnews_size
.RE
.RS
\-rc \-\-resetcounter
.RE
.RS
\-s  \-\-def_status_log
.RE
.RS
\-sg \-\-show_group
.RE
.RS
\-ssl \-\-use_ssl
.RE
.RS
\-w  \-\-wait_signal
.RE
.RS
\-x  \-\-no_chk_msgid
.RE
.RS
\-y  \-\-post_filter
.RE
.RS
\-z  \-\-no_dedupe
.RE
.RS
\-A  \-\-active
.RE
.RS
\-AL \-\-read_active
.RS
.RE
\-B   \-\-pre-batch
.RE
.RS
\-C  \-\-reconnect
.RE
.RS
\-D  \-\-debug
.RE
.RS
\-E  \-\-error_log
.RE
.RS
\-G  \-\-use_gui
.RE
.RS
\-H  \-\-no_history
.RE
.RS
\-HF \-\-history_file
.RE
.RS
\-K  \-\-killfile
.RE
.RS
\-L  \-\-kill_log_none
.RE
.RS
\-LS \-\-kill_log_short
.RE
.RS
\-LL \-\-kill_log_long
.RE
.RS
\-M  \-\-mode_reader
.RE
.RS
\-N  \-\-portnr
.RE
.RS
\-O \-\-skip_on_restart
.RE
.RS
\-P  \-\-password
.RE
.RS
\-Q  \-\-password_env
.RE
.RS
\-R  \-\-no_rescan
.RE
.RS
\-S  \-\-status_log
.RE
.RS
\-SSL \-\-local_use_ssl
.RS
\-T  \-\-timeout
.RE
.RS
\-U  \-\-userid
.RE
.RS
\-V  \-\-version
.RE
.RS
\-W  \-\-wait
.RE
.RS
\-X  \-\-no_xover
.RE
.RS
\-Z \-\-use_xover
.RE

.SH DESCRIPTION

.SH MODE 1 \- stdout mode
.RS
%suck
.RE
.RS
%suck myhost.com
.RE
.PP
Suck grabs news from an NNTP server and sends the articles to
stdout. Suck accepts as argument the name of an NNTP server or
if you don't give an argument it will take the environment variable
NNTPSERVER. You can redirect the articles to a file or compress them
on the fly like "suck server.domain | gzip \-9 > output.gz".
Now it's up to you what you do with the articles.  Maybe
you have the output already on your local machine because you
used a slip line or you still have to transfer the output to your
local machine.
.SH MODE 2 \- Multifile mode
.RS
%suck \-m
.RE
.RS
%suck myhost.com \-m
.RE
.PP
Suck grabs news from an NNTP server and stores each article in a
separate file.  They are stored in the directory specified in suck_config.h or
by the \-dm command line option.
.SH MODE 3 \- Batch mode
.RS
%suck myhost.com \-b[irlf] batchfile
.RE
.RS
or %suck myhost.com \-bp -hl localhost
.RE
.RS
or %suck myhost.com \-bP NR -hl localhost
.RE
.RS
%suck myhost.com \-b[irlf] batchfile
.RE
.PP
Suck will grab news articles from an NNTP server and store them
into files, one for each article (Multifile mode).  The location of the files
is based on the defines in suck_config.h and the command line \-dm.  
Once suck is done downloading the articles, it will build a batch file
which can be processed by either innxmit or rnews, or it will call lmove
to put the files directly into the news/group/number format.

\-bi \- build batch file for innxmit.  The articles are left intact,
and a batchfile is built with a one\-up listing of the full path of each article.
Then innxmit can be called:

.RS
%innxmit localhost batchfile
.RE

\-bl \- suck will call lmove to put the articles into
news/group/number format.  You must provide the name of the
configuration file on the command line.  The following arguments from suck 
are passed to lmove:

.RS
The configuration file name (the batchfile name provided with this option)
.RE
.RS
The directory specified for articles (-dm or built-in default).
.RE
.RS
The errorlog to log errors to (-e or -E), if provided on the command line.
.RE
.RS
The phrases file (-l), if provided on the command line.
.RE
.RS
The Debug option, if provided on the command line.
.RE

\-br \- build batch file for rnews.  The articles are
concatenated together, with the #!rnews size
article separator.  This can the be fed to rnews:

.RS
%rnews \-S localhost batchfile
.RE

\-r filesize  specify maximum batch file size for rnews.  This option
allows you to specify the maximum size of a batch file to be fed to rnews.
When this limit is reached, a new batch file is created AFTER I finish
writing the current article to the old batch file.  The second and
successive batch files get a 1 up sequence number attached to the
file name specified with the -br.  Note that since I have to finish
writing out the current article after reaching the limit, the
max file size is only approximate.

\-bf \- build a batch file for innfeed.  This batchfile contains the
MsgID and full path of each article.  The main difference between this
and the innxmit option is that the innfeed file is built as the articles
are downloaded, so that innfeed can be posting the articles, even while
more articles are downloaded.

\-bp \- This option tells suck to build a batch file, and post the articles
in that batchfile to the localhost (specified with the \-hl option).  This option
uses the IHAVE command to post all downloaded articles to the local host.
The batch file is called suck.post, and is put in the temporary directory (-dt).
It is deleted upon completion, as are the successfully posted articles.
If the article is not wanted by the server (usually because it already exists on
the server, or it is too old), the article is also deleted.  If other errors
occur, the article is NOT deleted.
With the following command line, you can download and post articles without
worrying if you are using INND or CNEWS.  

.RS
%suck news.server.com -bp -hl localhost -A -c
.RE

\-bP NR \- This option works identically to \-bp above, except instead of
waiting until all articles are downloaded, it will post them to the local
server after downloading NR of articles.

.RS
%suck news.server.com -bP 100 -hl localhost -A -c
.RE

.SH SUCK ARGUMENT FILE
.PP
If you specify @filename on the command line, suck will read from filename and
parse it for any arguments that you wish to pass to suck.  You specify the
same arguments in this file as you do on the command line.  The arguments
can be on one line, or spread out among more than one line.  You may also
use comments.  Comments begin with '#' and go to the end of a line.  All
command line arguments override arguments in the file.

.RS
# Sample Argument file
.RE
.RS
-bi batch # batch file option
.RE
.RS
-M	# use mode reader option
.RE

.SH SUCKNEWSRC
.PP
Suck looks for a file
.I sucknewsrc
to see what articles you want and
which you already received. The format of sucknewsrc is very simple. It
consists of one line for each newsgroup.  The line contains two or
three fields.

The first field is the name of the group.
 
The second field is the highest article number that was in the group
when that group was last downloaded.

The third field, which is optional, limits the number of articles which
can be downloaded at any given time.  If there are more articles than this
number, only the newest are downloaded.  If the third field is 0, then
no new messages are downloaded.  If the command line option \-lr is specified,
instead of downloading the newest articles, suck will download the oldest
articles instead.

The fields are separated by a space.

.RS
comp.os.linux.announce 1 [ 100 ]
.RE
.PP
When suck is finished, it creates the file suck.newrc which contains the
new sucknewsrc with the updated article numbers.
.PP
To add a new newsgroup, just stick it in sucknewsrc, with a
highest article number of \-1 (or any number less than 0).
Suck will then get the newest X number of messages for that newsgroup.
For example, a -100 would cause suck to download the newest 100
articles for that newsgroup.
.PP
To tell suck to skip a newsgroup, put a # as the first
character of a line.

.SH SUCKKILLFILE and SUCKXOVER
There are two types of killfiles supported in suck.  The first, via
the file suckkillfile, kills articles based on information in the
actual article header or body.   The second, via the file suckxover,
kills articles based on the information retreived via the NNTP command
XOVER.  They are implemented in two fundamentally different ways.  The
suckkillfile killing is done as the articles are downloaded, one at a
time.  The XOVER killing is done while suck is getting the list of articles
to download, and before a single article is downloaded.  You may use
either, none or both type of killfiles.

.SH SUCKKILLFILE and GROUP KEEP/KILLFILES
If
.I suckkillfile
exists,  the headers of  all articles will be scanned and the article downloaded or not, 
based on the parameters in the files.  If no logging option is specified (see the -L options
above), then the long logging option is used.
.PP
Comments lines are allowed in the killfiles.  A comment line has a "#" in the first position.
Everything on a comment line is ignored.
.PP
Here's how the whole keep/delete package works.  All articles are checked against the
master kill file (suckkillfile).  If an article is not killed by the master kill file,
then its group line is parsed.  If a group file exists for one of the groups then the
article is checked against that group file.  If it matches a keep file, then it is
kept, otherwise it is flagged for deletion.  If it matches a delete file, then it is
flagged for deletion, otherwise it is kept.  This is done for every group on the group line.
.PP
NOTES: With the exception of the USE_EXTENDED_REGEX parameter, none of these parameters are
passed from the master killfile to the individual group file.  Each killfile is separate
and independant.  Also, each search is case-insensitive unless specifically specified by starting the

CHANGELOG
CHANGES
=======
-- Changes from suck-4.3.2 -> Suck-4.3.3

- General code cleanup
- Applying patches from Debian, Fedora and openSUSE
- Add support for IPv6
- Manpage fixes
- Default path for history file is now /var/lib/news/history
- Default path for phrases file is now /etc/suck/phrases
- Install to DESTDIR
- Don't strip binaries

October 25, 2017 - Suck-4.3.3 released

-- Changes from suck-4.3.1 -> Suck-4.3.2

Wherever a hostname is specified on the command line, the form Host:Port
may be used to specify a port number.  If this form is used, the -N port
option will be ignored.

configure.in - add -with-perl-exe option to specify an
	alternate path to the perl exe.

suck.c, killfile.c - get_a_chunk(), get_chunk_mem()
	changed to handle lines beginning with .
	correctly in stdout mode.

suck.c, rpost.c - main() - error handling if host isn't
	specified on command line.

* - various patches from Debian folks
configure.in - check for configdata.h in inn include directory

both.c - sgetline() - changes to fix SSL bug with the select,
	connect_to_nntphost() - add code to handle
	host:port syntax.   This fix also allows you to
	specify a port number for the local host.

suck.c - to handle SIGINT in addition to SIGTERM,
	build_command() fix in potential buffer overflow

testhost.c - added -Q option, to allow user to specify
	NNTP authentiation via env variables.

March 28, 2003 - Suck-4.3.2 released

-- Changes from Suck-4.3.0 -> Suck-4.3.1

* as of this version, my e-mail address has changed
to bobyetman@sucknews.org

* 	- added stuff to autoconf to handle the Db,
	history, Perl & SSL stuff in the Makefile.

* both.h - added tests for socks.h

* rpost.c - added -Q option, to allow user
	to specify NNTP authentication via
	environment variables
         - do_batch() - add test to not append
	prefix if article name is an Inn token.

* suck.c - added -Q option, to allow user
	to specify NNTP authentication via
	environment variables

	- build_command() - changed sprintfs
	to snprintf() to fix potential overflow
	problems.

	-- get_one_article() - fix bug where
	plist would be null when doing authentication
	and there was only one article to download.

        - main() - fix bug where if reconnect after
	dedupe, and it failed, you would try to send
	the quit command to an invalid file descriptor.

Sep 24, 2002, Suck-4.3.1 released

-- Changes from Suck-4.2.5 -> Suck-4.3.0

* As of this version, my e-mail address has changed from
from bobyetman@worldnet.att.net to bobyetman@home.com

* 	changes to allow it to use SSL.

* rpost.c - added -i option, to ignore the
	readonly opening response and try
	to post anyway, since inn-2.3, when
	using authinfo, still sends 201 code
	meaning readonly.

	- do_authenticate() - added print
	statements to show start of, and
	successful completion of authentication.

* suck.c - get_one_article_kill() - add test
	for header_only mode when in stdout mode 
	to not download the body.

	- get_one_article(), get_articles() - added
	 -bP option, to post articles during download.

Sep 17, 2001 - Suck-4.3.0 released

-- Changes from Suck-4.2.4 -> Suck-4.2.5

* batch.c - post_one_msg() - fixed bug in
	testing for a long input line that
	caused us to reject any line that is
	exactly 13 char long.

* chhistory_db.c - various changes to allow it
	to compile with Inn-2.3 headers.

* rpost.c - do_article() - added test for second
	duplicate article string to handle DNews.

	- filter_batch_article() - added option for
	showing the file name as we upload it.

Dec 02, 2000 - Suck-4.2.5 released

-- Changes from Suck-4.2.3 -> Suck-4.2.4

*	various changes to allow compile under OS/2
	(look for ifdef EMX).  You'll need to edit the
	Makefile to get it to compile.

*both.c - true_str(), null_str() - moved from suck.c so
	rpost can use.

* killprg.c - Changes to include so will compile using 
	Perl5.6.  If you're using perl5.004 or older,
	you'll need to define OLD_PERL in the Makefile.

* rpost.c - Changes so will compile with perl5.6.0.
  
	do_article() - Added actual error message from
	remote server to Malfunction error message, to
	make it easier to figure out problem. 

* suck.c - do_one_group() - added code for resetcounter 
	option.  If remote end resets its article numbers
	the normal suck response is to ignore the group
	and reset the lastread counter to match the current
	high counter.  This option tells suck to reset
	its lastread counter to the low counter, effectively
	causing suck to get all articles for the group, and
	use the historydb to check for dupes.
	Added low_read option. This option is used in concert
	with the maxread option of the sucknewsrc.  Normally
	when there is a maxread in the sucknewsrc, suck will
	download the newest articles.  This option tells suck
	to download instead the oldest articles.
	
	do_articles() - added code for show_group option. 
	This option will add the name of the current group
	to the BPS display as you download the articles.
        
May 21, 2000 - Suck-4.2.4 released

-- Changes from Suck-4.2.2 -> Suck-4.2.3

* both.c - signal_block() - Changes to get rid of invalid 
	#ifdef POSIX_SIGNAL causing this routine never to
	get called.

* killfile.c. parse_killfile() - fixed bug in getting perl
        filter filename.  If you didn't specify debug, the
        nl was getting left on the name, causing errors.

* rpost.c - * - Major internal re-write to handle future
	expansion of filters/batch mode.  Changed many
	function's arg passing, so they only pass around
	myargs.  Re-wrote batch mode to make adding new
	filters easier.  Added rnews mode, for reading
	rnews files.  See man page for more details.

* suck.c - main(), sighandler() - changes from signal() to
	sigaction() so I use all POSIX signal handling stuff.

        - do_nodownload() - added, this routine allows you
        to specifiy Message-IDs to never download.  See the
        SUCKNODOWNLOAD section in the man page.

23 Jan 2000 - Suck-4.2.3 released


java/Suck.java
import java.awt.*;
import java.awt.event.*;
import java.io.*;
import java.lang.*;
import java.text.*;

public class Suck  {
    public static void main(String[] args) {
	/* first declare all our variables */
	Frame f; /* our master frame */
	Panel ctl_pan, stat_pan, errs_pan; /* this panel will always stay on the screen  */
	TextArea stats, errs; /* where we display the messages and errors */
	TextField ctl_count, ctl_speed; /* where we display count and BPS */
	Label stat_lab, err_lab, ctl_count_label, ctl_speed_label;
	Button exit_button; /* how we exit our program */
	Runtime rt; /* the runtime environment for our program */
	Process prcs; /* the actual program we run */
	InputStreamReader stdin, stderr; /* output from prg */
	BufferedReader in_br, err_br;    /* buffered output */
	Thread std_thread, err_thread;   /* threads to update TextAreas*/
	Menu Option; /* our option menu */
	MenuBar mb;  /* the menu bar */
	MenuItem menu_exit;

	/* My custom classes */
	SuckMenu menu_action;
	SuckUpdate std_upd, err_upd;    /* for our Threads */
	 
	/* sanity check our args */
	if(args.length == 0) {
	    System.out.println("No command to run, aborting\n");
	    System.exit(0);
	}

	/* set up our frame and size it */
	f = new Frame("suck");
	f.setSize(650,300);

	/* set up our panel for our status messages */
	stat_pan = new Panel();
	stats = new TextArea( 5,60 );
	stat_lab = new Label("Status Messages");
	stat_pan.add(stat_lab);	
	stat_pan.add(stats);
	    
	/* set up  our panel for our error messages */
	errs_pan = new Panel();
	errs = new TextArea( 5,60);
	err_lab = new Label("Error Messages");
	errs_pan.add(err_lab);
	errs_pan.add(errs);
	
	/* declare our panel for our controls */
	ctl_pan = new Panel();

	/* now set up our textfields for message count and BPS */
	ctl_count = new TextField("",10);
	ctl_count_label = new Label("Messages Left = ");
	ctl_speed = new TextField("",10);
	ctl_speed_label = new Label("Download Speed = ");
		
	/* make em non-editable by user */
	ctl_count.setEditable(false);
	ctl_speed.setEditable(false);

	/* add em to our frame */
	ctl_pan.add(ctl_count_label);
	ctl_pan.add(ctl_count);
	ctl_pan.add(ctl_speed_label);
	ctl_pan.add(ctl_speed);

	/* now add em to our master frame */
	f.add(ctl_pan,BorderLayout.NORTH);
	f.add(stat_pan,BorderLayout.CENTER);	
	f.add(errs_pan,BorderLayout.SOUTH);

	/* set up our menu bar */
	/* first, set up our menu item(s) */
	menu_exit = new MenuItem("Exit"); /* this is the word we display */
	menu_exit.setActionCommand("byebye"); /* this is the action */

	/* now add em to our menu */
	Option = new Menu("Options", true);
	Option.add(menu_exit);

	/* now set up our action */
	menu_action = new SuckMenu();
	Option.addActionListener(menu_action);
	
	/* now add the menu to our menu bar */
	mb = new MenuBar();
	mb.add(Option);
	f.setMenuBar(mb);

	/* all done setting up, show it */
	f.show();

	// now run our program 
	try {
	    rt = Runtime.getRuntime();
	    prcs = rt.exec(args);
	    
	    // get stdin first
	    stdin = new InputStreamReader(prcs.getInputStream());
	    in_br = new BufferedReader(stdin);
	    // now stderr
	    stderr = new InputStreamReader(prcs.getErrorStream());
	    err_br = new BufferedReader(stderr);

	    // Now set up and run our threads.
	    std_upd = new SuckUpdate();
	    err_upd = new SuckUpdate();
	
	    /* set up the stats one to parse message count and BPS */
	    std_upd.setUp(stats, in_br, ctl_count, ctl_speed);
	    /* error one doesn't parse message count and BPS */
	    err_upd.setUp(errs, err_br);

	    std_thread = new Thread(std_upd);
	    err_thread = new Thread(err_upd);

	    std_thread.start();
	    err_thread.start();

	} catch(IOException ioe) {
	    System.out.println(ioe);
	}
    }
}
// following is an class to handle the menu
class SuckMenu implements ActionListener {

    public void actionPerformed (ActionEvent e) {
	String s = e.getActionCommand();

	if("byebye".equals(s)) {
	    System.out.println("Exiting program.");
	    System.exit(0);
	}	
    }
}
// this class updates the status and message areas
class SuckUpdate implements Runnable {
    TextArea Text;   // area to update
    BufferedReader Data;  // where we get our data from
    TextField MsgCount, BPS;

    /* if only called with these two we won't parse messagecount and BPS */
    public void setUp(TextArea inp, BufferedReader br) {
	Text = inp;
	Data = br;
	MsgCount = null;
	BPS = null;
    }
    public void setUp(TextArea inp, BufferedReader br, TextField count, TextField bps) {
	Text = inp;
	Data = br;
	MsgCount = count;
	BPS = bps;
    }
    public void run() {
	String line;
	int count = 0;
	double bps = 0.0;
	Double dbl;
	int sep;
	DecimalFormat bpsf = new DecimalFormat( "#.0");

	try {
	    while((line = Data.readLine()) != null) {
		if(BPS != null && line.startsWith("---", 0)) {
		    /* handle our message count and BPS lines */
		    /* the lines are formatted as follows*/
		    /* ---message_nr+++ BPS */
		    /* BPS may not be there */
		    sep = line.indexOf("+++");
		    try {
			count = Integer.parseInt(line.substring(3,sep));
			if(sep+3 < line.length()) {
			    /* get the BPS which is a double, and there's no direct way to get to a string  */
			    dbl = Double.valueOf(line.substring(sep+3));
			    bps = dbl.doubleValue();
			}
		    } catch (NumberFormatException nfe) {
			System.out.println(line + ":" + nfe);
			System.out.println("sep = " + sep + " count =" + count + " bps=" +bps);
		    }
		    MsgCount.setText(String.valueOf(count));
		    BPS.setText(bpsf.format(bps) + " BPS");
		}
		else {
		    Text.append(line);
		    Text.append("\n");
		}
	    }
	    Data.close();
	}
	catch (IOException ioe) {
	    System.out.println(ioe);
	}
    }
}

